frontend/app/.server/auth/raoidc.service.ts

```diff
@@ -0,0 +1,238 @@
+
+    const authCode = new URL(request.url).searchParams.get('code');
+    const error = new URL(request.url).searchParams.get('error');
+    const state = new URL(request.url).searchParams.get('state');
+
+    if (error) {
+      throw new Error(`Unexpected error: ${error}`);
+    }
+
+    if (!authCode) {
+      throw new Error('Missing authorization code in response');
+    }
+
+    if (state !== expectedState) {
+      throw new Error(`CSRF error: incoming state [${state}] does not match expected state [${expectedState}]`);
+    }
+
+    const privateDecryptionKey = await generateCryptoKey(AUTH_JWT_PRIVATE_KEY, 'decrypt');
+    const privateSigningKey = await generateCryptoKey(AUTH_JWT_PRIVATE_KEY, 'sign');
+    const privateKeyId = generateJwkId(await subtle.exportKey('jwk', privateSigningKey));
+
+    const client: ClientMetadata = {
+      clientId: AUTH_RAOIDC_CLIENT_ID,
+      privateDecryptionKey: privateDecryptionKey,
+      privateSigningKey: privateSigningKey,
+      privateKeyId: privateKeyId,
+    };
+
+    const { jwkSet, serverMetadata } = await this.moizedFetchServerMetadata(AUTH_RAOIDC_BASE_URL, this.fetchFn);
+    const { accessToken, idToken } = await fetchAccessToken(serverMetadata, jwkSet, authCode, client, codeVerifier, redirectUri, this.fetchFn);
+    const userInfoToken = await fetchUserInfo(serverMetadata.userinfo_endpoint, jwkSet, accessToken, client, this.fetchFn);
+
+    return { accessToken, idToken, userInfoToken };
+  }
+
+  async handleSessionValidation({ request, session }: HandleSessionValidationArgs): Promise<void> {
+    this.log.debug('Performing RAOIDC session validation check');
+    const { AUTH_RAOIDC_BASE_URL, AUTH_RAOIDC_CLIENT_ID } = this.serverConfig;
+
+    const { pathname, searchParams } = new URL(request.url);
+    const returnTo = encodeURIComponent(`${pathname}?${searchParams}`);
+
+    if (!session.has('idToken') || !session.has('userInfoToken')) {
+      this.log.debug(`User has not authenticated, redirecting to /auth/login?returnto=${returnTo}`);
+      throw redirect(`/auth/login?returnto=${returnTo}`);
+    }
+
+    const idToken: IdToken = session.get('idToken');
+    const userInfoToken: UserinfoToken = session.get('userInfoToken');
+
+    if (userInfoToken.mocked) {
+      this.log.debug('Mocked user; skipping RAOIDC session validation');
+      return;
+    }
+
+    // idToken.sid is the RAOIDC session id
+    const sessionValid = await validateSession(AUTH_RAOIDC_BASE_URL, AUTH_RAOIDC_CLIENT_ID, idToken.sid, this.fetchFn);
+
+    if (!sessionValid) {
+      this.log.debug(`RAOIDC session has expired, redirecting to /auth/login?returnto=${returnTo}`);
+      throw redirect(`/auth/login?returnto=${returnTo}`);
+    }
+
+    this.log.debug('Authentication check passed');
+  }
+
+  private moizedFetchServerMetadata = moize(fetchServerMetadata, {
+    onCacheAdd: () => {
+      this.log.info('Creating new fetchServerMetadata memo');
+    },
+  });
```

Comment:
This should have a TTL because the contents of the discovery document could change.

Reply:
I will add a TODO.